core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -561,79 +560,62 @@ class GroupMetadataManager(val brokerId: Int,
+          info(s"Group $groupId transitioned to Dead in generation ${group.generationId}")
+          group.transitionTo(Dead)
         }
+        (expiredOffsets, group.is(Dead), group.generationId)
       }
-    }
-
-    for ((group, offsetsPartition, tombstones, numOffsetsExpired) <- result) {
-      val partitionOpt = replicaManager.getPartition(Topic.GroupMetadataTopicName, offsetsPartition)
-
-      partitionOpt.foreach { partition =>
-        val appendPartition = TopicAndPartition(Topic.GroupMetadataTopicName, offsetsPartition)
-        trace("Marked %d offsets in %s for deletion.".format(numOffsetsExpired, appendPartition))
 
-        try {
-          // do not need to require acks since even if the tombstone is lost,
-          // it will be appended again in the next purge cycle
-          partition.appendMessagesToLeader(new ByteBufferMessageSet(config.offsetsTopicCompressionCodec, tombstones: _*))
-          offsetsRemoved += numOffsetsExpired
-        }
-        catch {
-          case t: Throwable =>
-            error(s"Failed to write ${tombstones.size} tombstones for group ${group.groupId} to $appendPartition.", t)
-          // ignore and continue
-        }
+      val offsetsPartition = partitionFor(groupId)
+      getMessageFormatVersionAndTimestamp(offsetsPartition) match {
+        case Some((magicValue, timestamp)) =>
+          val partitionOpt = replicaManager.getPartition(Topic.GroupMetadataTopicName, offsetsPartition)
+          partitionOpt.foreach { partition =>
+            val appendPartition = TopicAndPartition(Topic.GroupMetadataTopicName, offsetsPartition)
+            val tombstones = expiredOffsets.map { case (topicPartition, offsetAndMetadata) =>
+              trace(s"Removing expired offset and metadata for $groupId, $topicPartition: $offsetAndMetadata")
+              val commitKey = GroupMetadataManager.offsetCommitKey(groupId, topicPartition.topic, topicPartition.partition)
+              new Message(bytes = null, key = commitKey, timestamp = timestamp, magicValue = magicValue)
+            }.toBuffer
+            trace(s"Marked ${expiredOffsets.size} offsets in $appendPartition for deletion.")
+
+            // We avoid writing the tombstone when the generationId is 0, since this group is only using
+            // Kafka for offset storage.
+            if (groupIsDead && groupMetadataCache.remove(groupId, group) && generation > 0) {
+              // Append the tombstone messages to the partition. It is okay if the replicas don't receive these (say,
+              // if we crash or leaders move) since the new leaders will still expire the consumers with heartbeat and
+              // retry removing this group.
+              tombstones += new Message(bytes = null, key = GroupMetadataManager.groupMetadataKey(group.groupId),
+                timestamp = timestamp, magicValue = magicValue)
+              trace(s"Marked group $groupId in $appendPartition for deletion")
```

Comment:
Nit: is it worth mentioning that it has been removed from the cache (we used to do that later, but we now do it here)?